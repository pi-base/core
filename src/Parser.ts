/// <reference path="Parser.d.ts" />
import remark from 'remark'
import remark2rehype from 'remark-rehype'
import visit from 'unist-util-visit'

// Adapted from
// * https://github.com/djm/remark-shortcodes/blob/master/index.js
// * https://using-remark.gatsbyjs.org/custom-components/
const delimitedParser = (type: string, start: string, stop: string) => {
  function parser(eat: Function, value: string, silent: boolean) {
    if (!value.startsWith(start)) {
      return
    }

    const stopPosition = value.indexOf(stop, start.length)
    if (stopPosition === -1) {
      return
    }

    if (silent) {
      return true
    }

    const tag = value.slice(0, stopPosition + stop.length)
    const inner = tag.slice(start.length, stopPosition)

    return eat(tag)({ type, [type]: inner })
  }

  parser.locator = (value: string, from: number) => value.indexOf(start, from)

  return parser
}

function pibase(this: any) {
  const parser = this.Parser.prototype

  parser.inlineTokenizers.citation = delimitedParser('citation', '{{', '}}')
  parser.inlineTokenizers.internalLink = delimitedParser(
    'internalLink',
    '{',
    '}'
  )
  parser.inlineTokenizers.inlineMathDollars = delimitedParser(
    'inlineMath',
    '$',
    '$'
  )
  parser.inlineTokenizers.inlineMathParens = delimitedParser(
    'inlineMath',
    '\\(',
    '\\)'
  )

  // It is important that the \(...\) tokenizer appears before the escape tokenizer
  //   but we might want to defer the others to later in the chain
  parser.inlineMethods.splice(
    parser.inlineMethods.indexOf('escape'),
    0,
    'citation',
    'internalLink',
    'inlineMathDollars',
    'inlineMathParens'
  )

  const transformer = (tree: any) => {
    visit(tree, 'citation', (node) => {
      node.data = {
        hName: 'citation',
        hProperties: {
          citation: node.citation,
        },
      }
    })
    visit(tree, 'inlineMath', (node) => {
      node.data = {
        hName: 'inlineMath',
        hProperties: {
          inline: true,
          formula: node.inlineMath,
        },
      }
    })
    visit(tree, 'internalLink', (node) => {
      node.data = {
        hName: 'internalLink',
        hProperties: {
          to: node.internalLink,
        },
      }
    })
  }

  return transformer
}

export default () => remark().use(pibase).use(remark2rehype)
