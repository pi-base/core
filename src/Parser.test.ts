import type { Node } from 'unist'
import visit from 'unist-util-visit'
import Parser from './Parser'

function parse(input: string) {
  const parser = Parser()
  return parser.runSync(parser.parse(input))
}

it('parses a citation', () => {
  const parsed = parse('{{doi:123}}')

  expect(deposition(parsed)).toEqual({
    type: 'root',
    children: [
      {
        type: 'paragraph',
        children: [
          {
            type: 'citation',
            citation: 'doi:123',
            data: {
              hName: 'citation',
              hProperties: {
                citation: 'doi:123',
              },
            },
          },
        ],
      },
    ],
  })
})

it('parses a complex example', () => {
  const example = `Inline math $2 + 2 = 4$ and display math $$2 + 2 = 4$$.

This is a list of links

* {S000123}
* {P000123}
* {T000123}
* {{doi:123}}
* {{mr:123}}
* {{wikipedia:123}}
* {{mathse:123}}
* {{mo:123}}`

  expect(parse(example)).toMatchSnapshot()
})

function deposition(tree: Node) {
  visit(tree, (node) => delete node['position'])
  return tree
}
