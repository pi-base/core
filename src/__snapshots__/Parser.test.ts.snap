// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`parses a complex example 1`] = `
Object {
  "children": Array [
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 1,
              "offset": 0,
            },
          },
          "type": "text",
          "value": "Inline math ",
        },
        Object {
          "data": Object {
            "hChildren": Array [
              Object {
                "type": "text",
                "value": "2 + 2 = 4",
              },
            ],
            "hName": "span",
            "hProperties": Object {
              "className": Array [
                "math-inline",
              ],
            },
          },
          "inlineMath": "2 + 2 = 4",
          "position": Position {
            "end": Object {
              "column": 24,
              "line": 1,
              "offset": 23,
            },
            "indent": Array [],
            "start": Object {
              "column": 13,
              "line": 1,
              "offset": 12,
            },
          },
          "type": "inlineMath",
        },
        Object {
          "position": Position {
            "end": Object {
              "column": 42,
              "line": 1,
              "offset": 41,
            },
            "indent": Array [],
            "start": Object {
              "column": 24,
              "line": 1,
              "offset": 23,
            },
          },
          "type": "text",
          "value": " and display math ",
        },
        Object {
          "blockMath": "2 + 2 = 4",
          "data": Object {
            "hChildren": Array [
              Object {
                "type": "text",
                "value": "2 + 2 = 4",
              },
            ],
            "hName": "span",
            "hProperties": Object {
              "className": Array [
                "math-display",
              ],
            },
          },
          "position": Position {
            "end": Object {
              "column": 55,
              "line": 1,
              "offset": 54,
            },
            "indent": Array [],
            "start": Object {
              "column": 42,
              "line": 1,
              "offset": 41,
            },
          },
          "type": "blockMath",
        },
        Object {
          "position": Position {
            "end": Object {
              "column": 56,
              "line": 1,
              "offset": 55,
            },
            "indent": Array [],
            "start": Object {
              "column": 55,
              "line": 1,
              "offset": 54,
            },
          },
          "type": "text",
          "value": ".",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 56,
          "line": 1,
          "offset": 55,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 1,
          "offset": 0,
        },
      },
      "type": "paragraph",
    },
    Object {
      "children": Array [
        Object {
          "position": Position {
            "end": Object {
              "column": 24,
              "line": 3,
              "offset": 80,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 3,
              "offset": 57,
            },
          },
          "type": "text",
          "value": "This is a list of links",
        },
      ],
      "position": Position {
        "end": Object {
          "column": 24,
          "line": 3,
          "offset": 80,
        },
        "indent": Array [],
        "start": Object {
          "column": 1,
          "line": 3,
          "offset": 57,
        },
      },
      "type": "paragraph",
    },
    Object {
      "children": Array [
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "hName": "internalLink",
                    "hProperties": Object {
                      "to": "S000123",
                    },
                  },
                  "internalLink": "S000123",
                  "position": Position {
                    "end": Object {
                      "column": 12,
                      "line": 5,
                      "offset": 93,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 5,
                      "offset": 84,
                    },
                  },
                  "type": "internalLink",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 12,
                  "line": 5,
                  "offset": 93,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 5,
                  "offset": 84,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 12,
              "line": 5,
              "offset": 93,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 5,
              "offset": 82,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "hName": "internalLink",
                    "hProperties": Object {
                      "to": "P000123",
                    },
                  },
                  "internalLink": "P000123",
                  "position": Position {
                    "end": Object {
                      "column": 12,
                      "line": 6,
                      "offset": 105,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 6,
                      "offset": 96,
                    },
                  },
                  "type": "internalLink",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 12,
                  "line": 6,
                  "offset": 105,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 6,
                  "offset": 96,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 12,
              "line": 6,
              "offset": 105,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 6,
              "offset": 94,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "data": Object {
                    "hName": "internalLink",
                    "hProperties": Object {
                      "to": "T000123",
                    },
                  },
                  "internalLink": "T000123",
                  "position": Position {
                    "end": Object {
                      "column": 12,
                      "line": 7,
                      "offset": 117,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 7,
                      "offset": 108,
                    },
                  },
                  "type": "internalLink",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 12,
                  "line": 7,
                  "offset": 117,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 7,
                  "offset": 108,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 12,
              "line": 7,
              "offset": 117,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 7,
              "offset": 106,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "citation": "doi:123",
                  "data": Object {
                    "hName": "citation",
                    "hProperties": Object {
                      "citation": "doi:123",
                    },
                  },
                  "position": Position {
                    "end": Object {
                      "column": 14,
                      "line": 8,
                      "offset": 131,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 8,
                      "offset": 120,
                    },
                  },
                  "type": "citation",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 14,
                  "line": 8,
                  "offset": 131,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 8,
                  "offset": 120,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 14,
              "line": 8,
              "offset": 131,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 8,
              "offset": 118,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "citation": "mr:123",
                  "data": Object {
                    "hName": "citation",
                    "hProperties": Object {
                      "citation": "mr:123",
                    },
                  },
                  "position": Position {
                    "end": Object {
                      "column": 13,
                      "line": 9,
                      "offset": 144,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 9,
                      "offset": 134,
                    },
                  },
                  "type": "citation",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 13,
                  "line": 9,
                  "offset": 144,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 9,
                  "offset": 134,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 13,
              "line": 9,
              "offset": 144,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 9,
              "offset": 132,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "citation": "wikipedia:123",
                  "data": Object {
                    "hName": "citation",
                    "hProperties": Object {
                      "citation": "wikipedia:123",
                    },
                  },
                  "position": Position {
                    "end": Object {
                      "column": 20,
                      "line": 10,
                      "offset": 164,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 10,
                      "offset": 147,
                    },
                  },
                  "type": "citation",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 20,
                  "line": 10,
                  "offset": 164,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 10,
                  "offset": 147,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 20,
              "line": 10,
              "offset": 164,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 10,
              "offset": 145,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "citation": "mathse:123",
                  "data": Object {
                    "hName": "citation",
                    "hProperties": Object {
                      "citation": "mathse:123",
                    },
                  },
                  "position": Position {
                    "end": Object {
                      "column": 17,
                      "line": 11,
                      "offset": 181,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 11,
                      "offset": 167,
                    },
                  },
                  "type": "citation",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 17,
                  "line": 11,
                  "offset": 181,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 11,
                  "offset": 167,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 17,
              "line": 11,
              "offset": 181,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 11,
              "offset": 165,
            },
          },
          "spread": false,
          "type": "listItem",
        },
        Object {
          "checked": null,
          "children": Array [
            Object {
              "children": Array [
                Object {
                  "citation": "mo:123",
                  "data": Object {
                    "hName": "citation",
                    "hProperties": Object {
                      "citation": "mo:123",
                    },
                  },
                  "position": Position {
                    "end": Object {
                      "column": 13,
                      "line": 12,
                      "offset": 194,
                    },
                    "indent": Array [],
                    "start": Object {
                      "column": 3,
                      "line": 12,
                      "offset": 184,
                    },
                  },
                  "type": "citation",
                },
              ],
              "position": Position {
                "end": Object {
                  "column": 13,
                  "line": 12,
                  "offset": 194,
                },
                "indent": Array [],
                "start": Object {
                  "column": 3,
                  "line": 12,
                  "offset": 184,
                },
              },
              "type": "paragraph",
            },
          ],
          "position": Position {
            "end": Object {
              "column": 13,
              "line": 12,
              "offset": 194,
            },
            "indent": Array [],
            "start": Object {
              "column": 1,
              "line": 12,
              "offset": 182,
            },
          },
          "spread": false,
          "type": "listItem",
        },
      ],
      "ordered": false,
      "position": Position {
        "end": Object {
          "column": 13,
          "line": 12,
          "offset": 194,
        },
        "indent": Array [
          1,
          1,
          1,
          1,
          1,
          1,
          1,
        ],
        "start": Object {
          "column": 1,
          "line": 5,
          "offset": 82,
        },
      },
      "spread": false,
      "start": null,
      "type": "list",
    },
  ],
  "position": Object {
    "end": Object {
      "column": 13,
      "line": 12,
      "offset": 194,
    },
    "start": Object {
      "column": 1,
      "line": 1,
      "offset": 0,
    },
  },
  "type": "root",
}
`;
